tsdb/chunks/head_chunks.go

```diff
@@ -0,0 +1,511 @@
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(w.buf[:], uint64(mint))
+	if err := w.write(w.buf[:8]); err != nil {
+		return 0, err
+	}
+
+	binary.BigEndian.PutUint64(w.buf[:], uint64(maxt))
+	if err := w.write(w.buf[:8]); err != nil {
+		return 0, err
+	}
+
+	// The reference is set to the segment index and the offset where
+	// the data starts for this chunk.
+	//
+	// The upper 4 bytes are for the segment index and
+	// the lower 4 bytes are for the segment offset where to start reading this chunk.
+	chunkRef = w.chunkRef(uint64(w.seq()), uint64(w.n))
+
+	w.buf[0] = byte(chk.Encoding())
+	if err := w.write(w.buf[:1]); err != nil {
+		return 0, err
+	}
+
+	n := binary.PutUvarint(w.buf[:], uint64(len(chk.Bytes())))
+	if err := w.write(w.buf[:n]); err != nil {
+		return 0, err
+	}
+
+	if err := w.write(chk.Bytes()); err != nil {
+		return 0, err
+	}
+
+	return chunkRef, w.wbuf.Flush()
+}
+
+func (w *HeadReadWriter) chunkRef(seq, offset uint64) (chunkRef uint64) {
+	return (seq << 32) | offset
+}
+
+func (w *HeadReadWriter) shouldCutSegment(chunkLength int) bool {
+	return w.n == 0 || // First segment
+		// TODO: tune this boolean, cutting a segment for only 1 chunk would be inefficient.
+		(time.Since(w.curFileStartTime) > w.segmentTime && w.n > HeadSegmentHeaderSize) || // Time duration reached for the existing file.
+		w.n+int64(chunkLength+27) >= MaxSegmentSize
+}
+
+func (w *HeadReadWriter) seq() int {
+	return w.curFileSequence
+}
+
+func (w *HeadReadWriter) write(b []byte) error {
+	n, err := w.wbuf.Write(b)
+	w.n += int64(n)
+	return err
+}
+
+func (w *HeadReadWriter) cut() (err error) {
+	// Sync current tail to disk and close.
+	if err := w.finalizeCurFile(); err != nil {
+		return err
+	}
+
+	n, f, seq, err := cutSegmentFile(w.dirFile, headChunksFormatV1, false, 0)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err != nil {
+			f.Close()
+		}
+	}()
+
+	w.n = int64(n)
+	oldSeq := w.curFileSequence
+	w.curFileSequence = seq
+
+	// Write current time in milliseconds for mint and 0 for maxt.
+	w.curFileStartTime = time.Now()
+	binary.BigEndian.PutUint64(w.buf[:], uint64(w.curFileStartTime.UnixNano()/1e6))
+	if _, err := f.Write(w.buf[:8]); err != nil {
+		return err
+	}
+	w.n += 8
+	binary.BigEndian.PutUint64(w.buf[:], 0)
+	if _, err := f.Write(w.buf[:8]); err != nil {
+		return err
+	}
+	w.n += 8
+
+	oldFile := w.curFile
+
+	w.curFile = f
+	if w.wbuf != nil {
+		w.wbuf.Reset(f)
+	} else {
+		w.wbuf = bufio.NewWriterSize(f, 1024)
+	}
+
+	if oldFile != nil {
+		// Open it again with the new size.
+		newTailFile, err := fileutil.OpenMmapFile(oldFile.Name())
+		if err != nil {
+			return err
+		}
+		w.bsMtx.Lock()
+		// Closing the last mmapped file.
+		if err := w.cs[oldSeq].Close(); err != nil {
+			w.bsMtx.Unlock()
+			return err
+		}
+		w.cs[oldSeq] = newTailFile
+		w.bs[oldSeq] = realByteSlice(newTailFile.Bytes())
+		w.bsMtx.Unlock()
+	}
+
+	mmapFile, err := fileutil.OpenMmapFileWithSize(f.Name(), int(MaxSegmentSize))
```

Comment:
Something is wrong with this `w.bsMtx`.

Imagine that here at this point `w.Close` is happening from another goroutine and only after that below steps are executed:

```
	w.bsMtx.Lock()
	w.cs[w.curFileSequence] = mmapFile
	w.bs[w.curFileSequence] = realByteSlice(mmapFile.Bytes())
	w.bsMtx.Unlock()
```
..  new opened mmap file to be not closed, leaking the file descriptor.

I think we should lock the whole `cut` method.

Reply:
Locking the entire method would create problems in the queries which access these chunks. I added another way to take care of this: lock the `wbufLock` lock used in `WriteChunk` (which calls `cut`) inside `Close` too, and additionally check for `Close` being called inside `WriteChunk` and exit early. Does this looks fine now?